Add notice on quantity change and update `wc/store/cart` to use thunks

opr opened the pull request:
# Description
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->
This PR adds a notice when the quantity of an item changes unexpectedly. These are mostly edge case scenarios, however there is one legit "regular" use-case. The following changes trigger the notice.
- Shopper has a bundled item in the Cart. They update the main item, which in turn causes the bundled item's quantity to update. In this situation, there is no other clue that the bundled item's quantity updated too, so a snackbar is shown.
- An extension updates the quantity of an item in the cart during `extensionCartUpdate`. The cart is returned and the notification is shown.
- [edge case] The shopper has the Cart open in two tabs. They update item A's quantity in the first tab and then go to the second tab. In this tab, the quantity of item A has not been updated yet. In this tab, they update item B's quantity. This causes the server to return the cart (with item A's updated quantity). A notice is shown to make sure the user is aware that the quantity of an item other than the one they touched was affected.
- [edge case] The shopper adds an item that does not have a minimum/maximum quantity, then loads the Cart block. The merchant then sets a minimum/maximum quantity and then the shopper sets the quantity of an item in the cart. The item's quantity updates and a notice is shown.

# Data store changes

In addition to the above, this PR updates the actions and resolvers on the `wc/store/cart` data store to use thunks. This was required because we need to dispatch actions to the `core/notices` data store, and previously the only place we could do this was in the `wc/store/cart` reducer. It is an anti-pattern to dispatch actions inside a reducer.

Because the flow of `receiveCart` now needs to dispatch actions, we needed to turn it into a thunk (or generator). Given that the other data stores in our project use thunks, I took this route.

I had to change `apiFetchWithHeaders` in `assets/js/data/shared-controls.ts` to export both a generator and an async function. This was required because the collections data store was using `apiFetchWithHeaders` in a synchronous generator, so couldn't await promises, but `apiFetchWithHeaders` function was changed to return a promise. Instead I abstracted the logic to `doApiFetchWithHeaders` which can be called by both the async function and the generator.

# TypeScript changes

To make this work correctly with TypeScript, I created some new types:

- **`ThunkReturnType`**
  - This type informs TS of the return type of an action retrieved from `dispatch` or `useDispatch` that is a thunk.
- **`Thunk`**
  - A shorthand way of writing a function that returns a function that returns _something_
- **`ResolveSelectFromMap`**
  - This function maps the resolvers defined in `resolvers.ts` to the resolvers available on a @wordpress/data store.

I also defined some specific to the `wc/store/cart` data store:

- **`CartResolveSelectFromMap`**
  - maps the cart store's action creators to the dispatch function passed to thunks.
- **`CartDispatchFromMap`**
  - maps the cart store's resolvers to the resolveSelect function passed to thunks.

Finally, to get thunks that return values working with TS, I needed to update the `DispatchFromMap` type.

The changes observed in cb10482fdcf3602fc2a891a27d319ee0fa6ab9f2 in plain English are: if the action creator extends a function that returns a generator, then its type is a Promise that returns a generator, otherwise, if it extends `Thunk` (a function that returns a function) then its type is the type returned by the function the action creator returns. If it extends neither of these, the type is void (plain action creator).

# New functions

Please see 162577f01c6cc2f65bf34452ac8c3890f1879084 where I added `notifyIfQuantityLimitsChanged`, `notifyIfQuantityChanged`, and `notifyQuantityChanges` and see 05527e570efdb454a5fcf43f4d4bf06e4d21f404 where I added `isWithinQuantityLimits`.

- `notifyQuantityChanges` is simply a wrapper which calls both `notifyIfQuantityChanged` and `notifyIfQuantityLimitsChanged`.
- `notifyIfQuantityLimitsChanged` checks the incoming cart and compares each item to the old cart. If the quantity limits has changed, and the old quantity is outside of the limits, and that item is not currently changing quantity, we can assert that the server unexpectedly changed the quantity limits and it will be auto-updated in the client.
- `notifyQuantityChanges` checks the incoming cart and compares each item to the old cart. If the quantity has changed of an item, and that item is not currently changing quantity, we can assert that the server changed it without the user requesting it, so a notification can be shown.
- `isWithinQuantityLimits` checks to see if a `CartItem`'s current quantity is within the constraints of its `quantity_limits`.

# Tests
I updated `assets/js/data/cart/test/resolvers.ts` to test thunks now, rather than generators.

 <!-- Reference any related issues or PRs here -->

Fixes TBC

<!-- Don't forget to update the title with something descriptive. -->
<!-- If you can, add the appropriate labels -->

#### Accessibility

<!-- If you've changed or added any interactions, check off the appropriate items below. You can delete any that don't apply. Use this space to elaborate on anything if needed. -->

- [x] I've tested using only a keyboard (no mouse)
- [x] I've tested using a screen reader
- [x] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [x] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

### Testing

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [x] Unit tests
  * [ ] E2E tests

#### Internal developer testing
<details>
  <summary>Expand to see instructions</summary>


<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

##### Prerequisites

1. Install WooCommerce Min/Max Quantities and activate it.
2. Install WooCommerce Points and rewards and activate it.
3. Add points to your account ( WooCommerce -> Points and Rewards)
4. Create a coupon on your store.
5. Add 3 or more items to your cart. Go to the Cart block. Open the Cart block again in a **new tab**

##### Case 1 - coupons

1. Open a new tab and go to the Cart block, modify the quantity of an item on your cart in tab 2.
2. Go to tab 1, apply the coupon to your cart. Verify it is added correctly and no errors occur. Verify the quantity of the item updated in the previous step updates correctly, and that you see a notice.
3. Go to tab 2 and update the quantity of an item.
6. Go back to tab 1 and remove the coupon again. Verify it is removed correctly and no errors occur. Verify the quantity of the item updated in the previous step updates correctly, and that you see a notice.
7. Go to tab 2 and update the quantity of an item.
8. Go back to tab 1 and try to apply a coupon that does not exist. Ensure you see an error about the coupon, and verify the quantity of the item updated in the previous step updates correctly, and that you see a notice.

##### Case 2 - regular quantity updates.

1. Go to tab 2 and update the quantity of an item.
2. Go to tab 1 and modify the quantity of a different item on your cart. Verify the quantity of the item updated in the previous step updates correctly, and that you see a notice.

##### Case 5 - removing items

1. Go to tab 2 and update the quantity of an item.
2. Go to tab 1 and remove a different item from your cart. Verify it removes correctly, and that the quantity of the item updated in the previous step updates correctly, and that you see a notice.

##### Case 6 - `extensionCartUpdate`

1. Go to tab 2 and update the quantity of an item.
2. Go to tab 1 and apply your points by using the WooCommerce Points and Rewards button.
3. Verify they are added correctly, and that the quantity of the item updated in step 1 updates correctly, and that you see a notice.

##### Case 7 - Min/Max quantities.

1. Open a new tab and go to the WP Dashboard. Edit one of the items in your cart. Set a minimum quantity on the item.
2. Without reloading, go to the original cart tab. Modify the quantity of an item.
3. Check the quantity of the item with the minimum quantity has been updated correctly, and a notice shows. (Note, please ignore the red notice at the top of the page for now)
4. Do the same with a max quantity.
5. Do the same with a "Group of" quantity (so, if your item in the Cart has a quantity of 10, set the "Group of" to 6.)

###### Case 8 - Address changes

1. Go to tab 2 and update the quantity of an item.
2. Go to tab 1 and update your address using the shipping calculator.
3. Verify they address updates correctly, and that the quantity of the item updated in step 1 updates correctly, and that you see a notice.
</details>

### User facing testing

1. Add a coupon to your store.
2. Add some items to your cart and go to the Cart block.
3. Test changing the quantities of some items.
5. Test applying the coupon, it should work and you should see the discount applied.
6. Try applying a nonexistent coupon, you should see an error.
7. Ensure removing an item from your cart works correctly.
8. Continue to checkout and ensure placing the order works OK.
9. Edit a page and add the All Products Block. Ensure it loads correctly in the editor and in the front-end.
10. Try adding products to your cart from the All Products Block.

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Performance Impact

Practically negligible performance degradation in the client while quantities are checked. Degrades linearly with number of items in the cart.

### Changelog

> Show notices to the shopper if an item in the cart's quantity is updated automatically.

opr committed:
Add receiveCart thunk

opr committed:
Add mapCartResponseToCart helper

opr committed:
Add getItemsPendingQuantityUpdate selector

opr committed:
Update cart resolvers to be thunks

opr committed:
Remove RECEIVE_CART action and replace with SET_CART_DATA

receiveCart will turn into a thunk.

opr committed:
Add notifyQuantityChanges functions

opr committed:
Remove receiveCart from action type definition, replace with setCartData

opr committed:
Move apiFetchWithHeaders out of controls

This will just be a normal function since we'll be updating actions to thunks which will use this instead of a control.

opr committed:
Include thunks in actions file

opr committed:
Update receiveCart action to setCartData

opr committed:
Update applyCoupon action to a thunk

opr committed:
Update useStoreCartCoupons to get action from correct place

opr committed:
Update StoreCartCoupon types

opr committed:
Add types for Thunk and ThunkReturnType in mapped-types

opr committed:
Change applyCoupon to a thunk

opr committed:
Get applyCoupon, removeCoupon, receiveApplyingCoupon from useDispatch

This is to separate the concerns of actions vs. selectors. Previously the actions were fetched during useSelect which is not a pattern we use anywhere else in the codebase. Since we updated the MapToDispatch type, we can now get correctly typed thunks from the data store.

opr committed:
Improve apiFetchWithHeaders typings

opr committed:
Convert removeCoupon from generator to thunk

opr committed:
Add applyCoupon and removeCoupon to CartAction type

opr committed:
Remove unused old-style type-def

opr committed:
Add receiveApplyingCoupon & receiveRemovingCoupon to StoreCartCoupon

opr committed:
Correct issues with StoreCartCoupon type

These were not intended to reflect the actions in data store, rather the functions offered by the useStoreCartCoupons hook.

opr committed:
Update applyExtensionCartUpdate to a thunk

opr committed:
Update addItemToCart to thunk

opr committed:
Add ResolveSelectFromMap type that works with thunks

opr committed:
Add CartDispatchFromMap and CartResolveSelectFromMap types

We can add this to all data stores to get them working with thunks properly.

opr committed:
Add docs and update generic name in ResolveSelectFromMap

opr committed:
Add correct types for thunk resolvers in cart data store

opr committed:
Update removeItemFromCart to thunk

opr committed:
Update apiFetchWithHeaders to use generic

opr committed:
Update selectShippingRate to thunk

opr committed:
Update resolver tests to test correct thunk functionality

opr committed:
Update updateCustomerData to thunk

opr committed:
Update reducer test to reflect new action name

opr committed:
Update comments on CartDispatchFromMap and CartResolveSelectFromMap

opr committed:
Add quantity_limits to preview cart

opr committed:
Make notices speak when shown

opr committed:
Remove copilot comment

opr committed:
Add isWithinQuantityLimits function

This is because we shouldn't show a notice if the quantity limits change, but the item's quantity is still OK.

opr committed:
Add tests for notifyQuantityChanges

opr committed:
Show notice when multiple_of is updated

opr committed:
Update test to test for multiple_of changes

opr commented on assets/js/data/cart/action-types.ts:
```diff
@@ -1,5 +1,5 @@
 export const ACTION_TYPES = {
-	RECEIVE_CART: 'RECEIVE_CART',
+	SET_CART_DATA: 'SET_CART_DATA',
```
the action that previously dispatched `RECEIVE_CART`  has become a thunk, and instead, eventually dispatches `SET_CART_DATA`. This is to preserve the name and notion of "receiving" a cart from the server, which works the same as before, but now performs additional side effects.

opr commented on assets/js/previews/cart.ts:
```diff
@@ -51,6 +51,12 @@ export const previewCart: CartResponse = {
 			backorders_allowed: false,
 			show_backorder_badge: false,
 			sold_individually: false,
+			quantity_limits: {
```
The preview cart did not have these set, so it was an incorrect representation of what would really be seen by the Cart block.

opr committed:
Remove empty export

opr committed:
Remove controls from cart data store

Not needed anymore since the exported value from the shared-controls file was empty.

opr committed:
Export a control and async function for apiFetchWithHeaders

This is required because async functions cannot be called from sync generators.

opr committed:
Use control version of apiFetchWithHeaders in the collections store

opr committed:
Improve comments and remove incorrect TypeScript

mikejolley commented on assets/js/data/cart/actions.ts:
```diff
@@ -11,33 +11,27 @@ import type {
+import { DispatchFromMap, ReturnOrGeneratorYieldUnion } from '../mapped-types';
+import { CartDispatchFromMap } from './index';
+
+// `Thunks are functions that can be dispatched, similar to actions creators
```
```suggestion
// Thunks are functions that can be dispatched, similar to actions creators
```

mikejolley commented on assets/js/data/cart/actions.ts:
```diff
@@ -11,33 +11,27 @@ import type {
+ * An action creator that dispatches the plain action responsible for setting the cart data in the store.
  *
- * @param {CartResponse} response
+ * @param {Cart} cart the parsed cart object. (Parsed into camelCase).
```
Should we remove docblocks rather than maintain them, since we have TS?

opr replied:
Good point, I'll undo this one and any others I did, but will create an issue for the cooldown to update the rest. Would add a lot of noise to this already large PR.

mikejolley commented on assets/js/data/cart/actions.ts:
```diff
@@ -208,29 +202,29 @@ export const triggerAddedToCartEvent = ( {
-		// If updated cart state was returned, also update that.
-		if ( error.data?.cart ) {
-			yield receiveCart( error.data.cart );
+export const applyExtensionCartUpdate =
```
Like this. Easier to understand.

mikejolley commented on assets/js/data/cart/actions.ts:
```diff
@@ -362,32 +354,32 @@ export function* addItemToCart(
-	}
-	yield itemIsPendingDelete( cartItemKey, false );
-}
+		dispatch.itemIsPendingDelete( cartItemKey, false );
```
You used `finally` in some earlier refactors similar to this one. Should that be done here, or should `finally` be removed? See `removeCoupon`.

mikejolley commented on assets/js/data/cart/actions.ts:
```diff
@@ -398,42 +390,40 @@ export function* removeItemFromCart(
+				dispatch.receiveCart( error.data.cart );
+			}
+		}
+		dispatch.itemIsPendingQuantity( cartItemKey, false );
```
You used `finally` in some earlier refactors similar to this one. Should that be done here, or should `finally` be removed? See `removeCoupon`.

opr replied:
I've updated cases where it made sense in f237ab43bd1ebff99041d92b6aad3ca4c6b63a69. In 2e4a55254490a9919a0578274dcdbd985c820195 I added it to `removeItemFromCart` and `changeCartItemQuantity` because even if these actions error, we should still set the state to reflect that they're no longer removing or updating.

mikejolley commented on assets/js/data/cart/actions.ts:
```diff
@@ -442,38 +432,37 @@ export function* changeCartItemQuantity(
-	yield shippingRatesBeingSelected( false );
-	return true;
-}
+		dispatch.shippingRatesBeingSelected( false );
```
You used `finally` in some earlier refactors similar to this one. Should that be done here, or should `finally` be removed? See `removeCoupon`.

mikejolley commented on assets/js/data/cart/notify-quantity-changes.ts:
```diff
@@ -0,0 +1,171 @@
+				sprintf(
+					/* translators: %1$s is the name of the item, %2$d is the quantity of the item. %3$d is a number that the quantity must be a multiple of. */
+					__(
+						'The quantity of "%1$s" has been changed to %2$d. This product must be purchased in groups of %3$d.',
```
👍🏻

mikejolley replied:
I ran a few of these strings though Grammarly and it suggested switching to active voice. e.g. 

`The quantity of "test" has changed to 2. You must purchase this product in groups of 3.`

Worth using?

opr replied:
Nice, thanks for doing that. @elizaan36 could you look at the strings in this file and provide different wording if necessary?

The messages we will show are:

- The quantity of "Beanie" was changed to 2.
- The quantity of "Hoodie" was decreased to 5. This is the maximum allowed quantity.
- The quantity of "Album" was increased to 10. This is the minimum required quantity.
- The quantity of "Sunglasses" was changed to 6. You must purchase this product in groups of 3.

mikejolley commented on assets/js/data/shared-controls.ts:
```diff
@@ -114,84 +102,115 @@ const batchFetch = async ( request: APIFetchOptions ) => {
+		if (
+			! options.method ||
+			options.method === 'GET' ||
+			isWpVersion( '5.6', '<' )
```
Our min supported version exceeds this now.

mikejolley commented on assets/js/previews/cart.ts:
```diff
@@ -51,6 +51,12 @@ export const previewCart: CartResponse = {
 			sold_individually: false,
+			quantity_limits: {
+				minimum: 1,
+				maximum: 99,
```
Noting that there is one case in the API where maximum is `null`, rather than integer. Do you want to fix that in this PR too?

opr replied:
I changed it so that in that case, instead of maximum getting set to `null` it now gets set to `9999`

mikejolley reviewed (commented):
This is looking good! I added some feedback.

I did come across a test case that had unexpected results and may be related.

1. Tab 1: Remove item
2. Tab 2: Edit item
3. No notice about the removed item.

And test case 2:

1. Tab 1: Remove item from cart
2. Tab2: Edit the qty of that same item in the cart.
3. No visual errors/changes.

Should we adjust this?

opr committed:
Update assets/js/data/cart/actions.ts

Co-authored-by: Mike Jolley <mike.jolley@me.com>

opr committed:
Update ResolveSelectFromMap to include selectors too

opr committed:
Update TS in actions

opr committed:
Use finally to remove duplicate code

opr committed:
remove item pending delete/qty update after action runs in all cases

This will also reset the state when the request to remove it/change quantity errors

opr committed:
Remove unnecessary type from param.

Not needed because we have TS now. The description can stay though, it is useful.

opr committed:
Update snackbar wording to use active voice

elizaan36 commented:
@opr The strings look good to me.

alexflorisca commented on assets/js/data/cart/notify-quantity-changes.ts:
```diff
@@ -0,0 +1,171 @@
+
+const isWithinQuantityLimits = ( cartItem: CartItem ) => {
+	return (
+		cartItem.quantity >= cartItem.quantity_limits.minimum &&
```
Will the quantity limits always be defined here? Is it possible for the `.minium and `.maxium` props to be undefined?

opr replied:
Based on what comes back from the API, no. There was a case where `maximum` could be `null` but I changed that in  69f9570138254b3b0a255829a067e7b31452189f

opr committed:
Remove old WP version check

opr committed:
Set max quantity to high number instead of null

This would only happen in a niche case, and would require several TS changes to fix, so it's better to set it as a number here. 9999 should be high enough, and is the default quantity limit set below in get_product_quantity_limit

alexflorisca commented on assets/js/data/cart/test/notify-quantity-changes.ts:
```diff
@@ -0,0 +1,120 @@
+	const oldCart = mapKeys(
+		cloneDeep< CartResponse >( previewCart ),
+		( _, key ) => camelCase( key )
+	) as unknown as Cart;
```
Why are we using `as` twice with different values here?

opr replied:
Without the `as unknown` it complains and says the types don't match.

```
Type 'Dictionary<number | boolean | CartResponseCouponItem[] | CartResponseShippingRate[] | CartResponseShippingAddress | ... 7 more ... | ExtensionsData>' is missing the following properties from type 'Cart': coupons, shippingRates, shippingAddress, billingAddress, and 12 more
```

but we know they do because we know a `CartResponse` with camelCase keys is equivalent to a `Cart`.

alexflorisca commented on assets/js/data/cart/test/resolvers.ts:
```diff
@@ -0,0 +1,54 @@
+/**
```
Nice one for converting this to ts! 🎉

opr committed:
Set code on woocommerce_rest_cart_invalid_key to 409

This is so the cart is returned in the response, so the client can update.

opr committed:
Fix typo in comment and add CartSelectFromMap

opr committed:
Remove unnecessary docblock

opr committed:
Add getItemsPendingDelete selector

This is needed so we can show a notice for items that are unexpectedly removed from the cart. We need to know which ones are pending delete so we can skip showing the notice for them.

alexflorisca reviewed (commented):
Nice work, glad we're changing the controls to thunks in cart to align better with other data stores. Left some inline comments and I also noticed that when adding products to cart from the all products block, the mini cart doesn't update. Probably unrelated to this PR though.

opr committed:
Add type for notifyQuantityChanges args and change args to object

opr committed:
Add notifyIfRemoved function

This will check items that have been removed and show a notice for them.

opr committed:
Fix TS in receiveCart & pass itemsPendingDelete to notifyQuantiyChanges

opr committed:
Update wording on removal notice

opr committed:
Update types for notifyQuantityChanges args

opr committed:
Update tests to reflect new wording and args being an object

opr committed:
Check item is truth before running comparison of keys

opr committed:
Update tests for unexpectedly and expectedly removed items

opr committed:
Ignore print_r to satisfy phpcs

opr committed:
Update PHP tests to reflect correct response code when deleting items

opr commented:
Thanks for the review @alexflorisca

> I also noticed that when adding products to cart from the all products block, the mini cart doesn't update.

I just checked on trunk, looks like it's happening there too.

opr commented:
@mikejolley and @alexflorisca 

I have made some more changes:

- Add `notifyIfRemoved` function that checks for (unexpectedly) removed items.
- Update the response code when trying to remove an invalid item to 409 ([Conflict](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/409), so the client can resolve the differences)
- Updates to TS and removed unnecessary doc blocks
  - Added `CartSelectFromMap` type, which is similar to `CartResolveSelectFromMap` but does not include resolvers.
- Update JS unit tests to test for notification on unexpected item removal
- Update a PHP test to reflect new error response when invalid items are in the cart.
- Add `getItemsPendingDelete` selector in the cart data store. Necessary so we can retrieve the items slated for removal when checking if notices should be shown.

Mike please give it another go with the two cases you found while testing. Cheers
-

mikejolley reviewed (approved):
Excellent, test cases pass. Really good job!

alexflorisca commented on src/StoreApi/Utilities/QuantityLimits.php:
```diff
@@ -24,7 +24,7 @@ public function get_cart_item_quantity_limits( $cart_item ) {
 			return [
 				'minimum'     => 1,
-				'maximum'     => null,
+				'maximum'     => 9999,
```
Is this big enough? Thinking about wholesale shops for example that might sell in large quantities. Maybe we add a few more 9s?

opr replied:
In https://github.com/woocommerce/woocommerce-blocks/pull/7938/files/6c02026160322c51880d1e7618c71bf478282359#diff-1fa5b0bd80112a4242c903b56beb83d233673e7742fb0616650c782713799cf0R143 the limit is also 9999, and the line your comment refers to only happens in an edge-case, so I think it's OK :)

alexflorisca reviewed (commented):
Looks great, just one thing about the maximum quantity

opr committed:
Remove unnecessary controls and dispatch events directly from thunk

alexflorisca reviewed (approved):
Fine work sir
